# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [35]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
from sklearn import datasets

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [36]:
path=r"C:\Users\chris\Springboard_local\6_ApplyDSM\data"
os.chdir(path) 
os.getcwd()
df = pd.read_csv('step3_output.csv')
df.head()

Name       state  summit_elev  vertical_drop  \
0           Aspen / Snowmass    Colorado        12510           4406   
1         Deer Valley Resort        Utah         9570           3000   
2     Copper Mountain Resort    Colorado        12313           2738   
3  Mammoth Mountain Ski Area  California        11053           3100   
4                  Telluride    Colorado        13150           4425   

   base_elev  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0       8104      3          0          1         15     4  ...   
1       6570      1          0          0         13     0  ...   
2       9712      1          0          2          4     0  ...   
3       7953      3          0          2          9     1  ...   
4       8725      2          0          0          6     1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             5517.0             658        138.000000       72.0   
1             2026.0             660        115.103943       39.0   
2             2527.0             364        164.000000       47.0   
3             3500.0             700        243.000000       66.0   
4             2000.0             220        131.000000       47.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            300.0         179.0         179.0         138.000000   
1            300.0         169.0         169.0         120.053004   
2            300.0         158.0         158.0         164.000000   
3            400.0         149.0         159.0         120.053004   
4            280.0         139.0         139.0         137.000000   

   NightSkiing_ac  clusters  
0               0         1  
1               0         1  
2               0         1  
3               0         1  
4               0         1  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [37]:
df1 = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
df1.head()
#df_test is a dataframe that is being used to troubleshoot model 1
df_test = df.drop(['Name', 'state', 'clusters'], axis=1)
df_test.head()

summit_elev  vertical_drop  base_elev  trams  fastEight  fastSixes  \
0        12510           4406       8104      3          0          1   
1         9570           3000       6570      1          0          0   
2        12313           2738       9712      1          0          2   
3        11053           3100       7953      3          0          2   
4        13150           4425       8725      2          0          0   

   fastQuads  quad  triple  double  ...  LongestRun_mi  SkiableTerrain_ac  \
0         15     4       3       5  ...            5.3             5517.0   
1         13     0       5       2  ...            2.8             2026.0   
2          4     0       4       4  ...            1.7             2527.0   
3          9     1       6       4  ...            3.0             3500.0   
4          6     1       2       2  ...            4.6             2000.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0             658        138.000000       72.0            300.0         179.0   
1             660        115.103943       39.0            300.0         169.0   
2             364        164.000000       47.0            300.0         158.0   
3             700        243.000000       66.0            400.0         149.0   
4             220        131.000000       47.0            280.0         139.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  
0         179.0         138.000000               0  
1         169.0         120.053004               0  
2         158.0         164.000000               0  
3         159.0         120.053004               0  
4         139.0         137.000000               0  

[5 rows x 24 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [38]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df1.drop(['Name','AdultWeekend','clusters'], axis=1)
#X = df_test
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1['AdultWeekend']
#y = df_test['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [39]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [40]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [41]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [42]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [43]:
from sklearn.metrics import explained_variance_score, mean_absolute_error
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

-9.723233394532206e+22
1531263335330.3064


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [44]:
print(lm.intercept_)

-385135929729.93646


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [45]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
dfco = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfco.sort_values(by='Coefficient', ascending=False, inplace=True)
dfco.head(10)

Coefficient
total_chairs  3.439153e+13
fastQuads     1.303791e+13
New York      1.270949e+13
surface       1.221554e+13
Michigan      1.199430e+13
double        1.076479e+13
Colorado      1.056768e+13
California    1.034146e+13
Pennsylvania  9.868469e+12
triple        9.602930e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [46]:
X = df.drop(['state', 'AdultWeekend', 'Name', 'clusters'], axis=1)
y = df['AdultWeekend'] 
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [47]:
#all second model set
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.9077050692438466
5.9711518709331735


In [48]:
dfco = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfco.sort_values(by='Coefficient', ascending=False, inplace=True)
dfco.head(10)

Coefficient
AdultWeekday         19.571534
summit_elev           2.228855
vertical_drop         2.111471
daysOpenLastYear      1.751777
trams                 1.409968
quad                  1.338578
averageSnowfall       1.296570
triple                1.219782
surface               1.154373
projectedDaysOpen     0.996351

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [49]:
X = df.drop(['Name','state', 'AdultWeekend', 'clusters', 'summit_elev', 'base_elev'], axis=1)
y = df['AdultWeekend'] 
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [50]:
#all third model set
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.90983692043986
5.92806157134282


In [51]:
dfco = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
dfco.sort_values(by='Coefficient', ascending=False, inplace=True)
dfco.head(10)

Coefficient
AdultWeekday        19.564948
averageSnowfall      2.153025
daysOpenLastYear     1.579321
quad                 1.419361
surface              1.113669
trams                1.111913
vertical_drop        1.082585
fastQuads            1.081036
triple               1.070964
Snow Making_ac       0.879949

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | NA |NA  |-|
| Model 2. | 0.9077|5.9711 |'state'|
| Model 3. | 0.9098 |5.9280 |'state','summit_elev','base_elev'|

Model Selection:

In [52]:
#Model 3 would be the best fit